contracts/exchange/mixins/DcvSale.sol

```diff
@@ -0,0 +1,243 @@
+pragma solidity 0.8.19;
+
+import "@openzeppelin/contracts-upgradeable/security/ReentrancyGuardUpgradeable.sol";
+import "@openzeppelin/contracts/token/ERC721/IERC721.sol";
+import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+
+import "./DcvExchangeCore.sol";
+import "./DcvFundsDistributor.sol";
+import "./DcvEIP712.sol";
+import "./DcvNonceManager.sol";
+
+abstract contract DcvSale is
+    ReentrancyGuardUpgradeable,
+    DcvExchangeCore,
+    DcvFundsDistributor,
+    DcvEIP712,
+    DcvNonceManager
+{
+    using ECDSA for bytes32;
+
+    bytes32 private constant SALE_TYPEHASH =
+        keccak256(
+            "Sale(uint256 orderNonce,address nftContract,uint256 tokenId,uint256 price)"
+        );
+    bytes32 private constant SALE_WITH_ROYALTY_TYPEHASH =
+        keccak256(
+            "SaleWithRoyalty(uint256 orderNonce,address nftContract,uint256 tokenId,uint256 price,address royaltyRecipient,uint256 royaltyPercentage)"
+        );
+    bytes32 private constant ROYALTY_PARAMETERS_TYPEHASH =
+        keccak256(
+            "RoyaltyParameters(address nftContract,uint256 tokenId,address royaltyRecipient,uint256 royaltyPercentage)"
+        );
+
+    struct Sale {
+        address seller;
+        uint256 orderNonce;
+        address nftContract;
+        uint256 tokenId;
+        uint256 price;
+    }
+
+    struct SaleWithRoyalty {
+        address seller;
+        uint256 orderNonce;
+        address nftContract;
+        uint256 tokenId;
+        uint256 price;
+        address royaltyRecipient;
+        uint256 royaltyPercentage;
+    }
+
+    event SaleSuccess(
+        address indexed nftContract,
+        uint256 indexed tokenId,
+        address indexed seller,
+        address buyer,
+        uint256 price,
+        uint256 platformFee
+    );
+
+    function buyFromSale(
+        bytes calldata signature,
+        Sale calldata sale
+    ) external payable nonReentrant {
+        if (sale.seller == msg.sender) revert InvalidCaller();
+
+        _verifySaleSignature(signature, sale);
+
+        _requireSufficientEtherSupplied(sale.price);
+
+        _invalidateNonce(sale.seller, sale.orderNonce);
+
+        (
+            uint256 platformFee,
+            address royaltyRecipient,
+            uint256 royaltyFee
+        ) = _calculateFees(sale.nftContract, sale.tokenId, sale.price);
+
+        _finalizeSale(
+            sale.nftContract,
+            sale.tokenId,
+            sale.seller,
+            sale.price,
+            royaltyRecipient,
+            platformFee,
+            royaltyFee
+        );
+    }
+
+    function buyFromSaleWithRoyalty(
+        bytes calldata sellerSignature,
+        bytes calldata royaltySignature,
+        SaleWithRoyalty calldata sale
+    ) external payable nonReentrant {
+        if (sale.seller == msg.sender) revert InvalidCaller();
+        if (sale.seller != sale.royaltyRecipient) {
+            revert UnauthorizedRoyaltyChange();
+        }
```

Comment:
Sta nam je ovo?

Reply:
Cilj je da postignemo da se royalty parametri mogu minjat samo ako token prodaje originalni minter, i onda je royaltyRecipient upravo taj originalni minter. To jest želimo da potpisani royalty recipient od strane DCV teama za taj token bude jednak selleru tog tokena, posto DCV team potpisuje samo originalnog mintera kao royaltyRecipient nekog tokena, iz toga proizlazi da seller mora biti originalni minter.

Ispod ovoga se provjerava da je taj royaltyRecipient zaista potpisan od DCV teama, i da je seller to sve skupa potpisa.

Mislim da bi se bez te provjere (na liniji 97) omogucila potencijalna zloupotreba potpisanih royalty parametara. Ako DCV team na zahtjev originalnog mintera potpise vise royalty parametara npr.: (royaltyRecipient=originalni minter, royaltyPercentage=5%) i (royaltyRecipient=originalni minter, royaltyPercentage=0%). Recimo da je trenutno postavljeno 5%, onda bi sekundarni seller mogao prilozit validni potpis royalty parametara di je percentage 0%, i onda bi se to postavilo bez dopustenja originalnog mintera. To jest royalty parametri bi se teoretski mogli minjat ako seller nije originalni minter.